Sources_v3/StreamChat/ChatClient.swift

```diff
@@ -271,54 +264,214 @@ public class _ChatClient<ExtraData: ExtraDataTypes> {
     ///
     init(
         config: ChatClientConfig,
+        tokenProvider: _TokenProvider<ExtraData>,
         workerBuilders: [WorkerBuilder],
         eventWorkerBuilders: [EventWorkerBuilder],
         environment: Environment
     ) {
+        log.assert(Thread.isMainThread, "ChatClient.init must be called on main thread.")
+
```

Comment:
Is this really a condition we need?

Reply:
That's because of
```
currentUserId = databaseContainer.viewContext.currentUser()?.user.id
```
Checking the thread allows us to load the current user id synchronously since we are sure the thread is `main` we can access `viewContext` safely. Without this assertation, we have to load the current user via bg context asynchronously.

Why would some need to instantiate **ChatClient** on the background thread? 🤔

Reply:
It's quite a common practice to speed up app launch to move dependency instantiation you don't need immediately to a background thread. What about moving the handling of this situation to our side with:
```swift
        var existingUserId: UserId?
        
        if Thread.isMainThread {
            existingUserId = databaseContainer.viewContext.currentUser()?.user.id
        } else {
            databaseContainer.viewContext.performAndWait { [context = databaseContainer.viewContext] in
                existingUserId = context.currentUser()?.user.id
            }
        }
        
        print(existingUserId)
```